File: src/programmers/입국심사.java
package src.programmers;

import java.util.*;

class 입국심사 {
    public long solution(int n, int[] times) {
        return findMinTimeAllPersonChecked(n, times);
    }

    private long findMinTimeAllPersonChecked(int n, int[] times) {
        Arrays.sort(times);
        long left = 0;
        // 최대로 걸릴 수 있는 시간
        long right = (long) times[times.length - 1] * n;

        while (left <= right) {
            long mid = (left + right) / 2;
            // 해당 시간안에 모두 확인가능 -> 더 작은 시간안에 가능한지 확인
            if (countCheckedPerson(times, mid) < n) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return right + 1;
    }

    private long countCheckedPerson(int[] times, long endTime) {
        long count = 0;
        for (int time : times) {
            count += endTime / time;
        }
        return count;
    }
}

File: src/boj/october/Problem3079/입국심사.kt
package src.boj.october.Problem3079

private lateinit var times: List<Int>
private var M = 0L
fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    M = m.toLong()
    times = List(n) { readLine().toInt() }.sorted()
    search()
}

private fun search() {
    var left = 0L
    var right = M * times.last()

    while (left <= right) {
        val mid = (left + right) / 2
        if (count(mid)) {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }
    println(right + 1)
}

private fun count(endTime: Long): Boolean {
    var check = 0L
    for (time in times) {
        check += (endTime / time)
        if (M <= check) return true
    }
    return false
}
